File: VB.NET/CrossCall/CrossCall.vb
Imports System.Text
Imports System.Threading
Imports ApiHubTool
Imports ApiHubTool.Bindings

Module CrossCall

    Private _running As Boolean = True

    ' ========== 远程调用封装 ==========

    Private Function Add__(a As Integer, b As Integer) As Integer
        Dim param = API.API_Create_DataHnd("add")
        If Not param.IsValid Then Return 0

        API.WriteInt32(param, a)
        API.WriteInt32(param, b)

        Dim result = API.API_Call("demo", param, 15000)   ' 15秒超时
        API.API_Free_DataHnd(param)

        If Not result.IsValid OrElse API.API_GetSize(result) = 0 Then
            If result.IsValid Then API.API_Free_DataHnd(result)
            Return 0
        End If

        Dim ret As Integer
        If Not API.ReadInt32(result, ret) Then ret = 0
        API.API_Free_DataHnd(result)
        Return ret
    End Function

    Private Function InvSeri__() As String
        Dim param = API.API_Create_DataHnd("inv_seri")
        If Not param.IsValid Then Return "创建句柄失败"

        ' 写入 Pascal 约定的数据
        API.WriteUInt8(param, 200)
        API.WriteUInt16(param, &H10)
        API.WriteUInt32(param, &H2F)
        API.WriteUInt64(param, &H3F)
        API.WriteStringNullTerminated(param, "hello world")
        API.WriteSingle(param, 3.14F)

        Dim result = API.API_Call("demo", param, 15000)
        API.API_Free_DataHnd(param)

        If Not result.IsValid OrElse API.API_GetSize(result) = 0 Then
            If result.IsValid Then API.API_Free_DataHnd(result)
            Return "调用超时或失败"
        End If

        ' 逆序读取
        Dim f As Single
        Dim s As String
        Dim u64 As ULong
        Dim c As UInteger
        Dim w As UShort
        Dim b As Byte

        If Not API.ReadSingle(result, f) Then
            API.API_Free_DataHnd(result)
            Return "读取失败"
        End If
        If Not API.ReadStringNullTerminated(result, s) Then
            API.API_Free_DataHnd(result)
            Return "读取失败"
        End If
        If Not API.ReadUInt64(result, u64) Then Return "读取失败"
        If Not API.ReadUInt32(result, c) Then Return "读取失败"
        If Not API.ReadUInt16(result, w) Then Return "读取失败"
        If Not API.ReadUInt8(result, b) Then Return "读取失败"

        API.API_Free_DataHnd(result)

        Return $"接收数据序 [{b}, {w}, {c}, {u64}, ""{s}"", {f:F2}] = 发送数据序 [{f:F2}, ""{s}"", {u64}, {c}, {w}, {b}]"
    End Function

    ' ========== 工作线程 ==========

    Private Sub DoCompute()
        Dim startTime = DateTime.Now
        Dim duration = TimeSpan.FromSeconds(10)
        Dim rand As New Random()

        Console.WriteLine("仿真计算启动（可多开），持续 10 秒...")

        While _running AndAlso (DateTime.Now - startTime) < duration
            If rand.Next(2) = 0 Then
                Dim a = rand.Next(1, Integer.MaxValue)
                Dim b = rand.Next(1, Integer.MaxValue)
                Dim result = Add__(a, b)
                Dim remaining = (duration - (DateTime.Now - startTime)).TotalSeconds
                If result <> 0 Then
                    Console.WriteLine($"[Call] 计算 ""a({a})+b({b})"" = 计算结果 {result} ({remaining:F2}秒以后退出)")
                Else
                    Console.WriteLine($"[Call] add({a}, {b}) 返回 0（重试后仍失败） ({remaining:F2}秒退出)")
                End If
            Else
                Dim status = InvSeri__()
                Dim remaining = (duration - (DateTime.Now - startTime)).TotalSeconds
                Console.WriteLine($"[Call] {status} ({remaining:F2}秒退出)")
            End If
            Thread.Sleep(1)
        End While

        Console.WriteLine("工作线程结束")
    End Sub

    Sub Main()
        Console.WriteLine("=== VB.NET CrossCall (Concurrent Client) ===")

        API.API_Reset_Prepare()
        API.API_Prepare_Client("ipc:cross", AppHnd.Null)

        If API.API_Prepare_Done() <> 1 Then
            Console.WriteLine("连接服务注册中心失败。请查看控制台输出。")
            API.API_shutdown()
            Return
        End If

        Console.WriteLine("已连接到 ipc:cross")

        ' 启动工作线程
        Dim computeThread As New Thread(AddressOf DoCompute)
        computeThread.Start()

        ' 等待用户按 Enter 提前退出
        Console.WriteLine("按 Enter 提前终止...")
        Console.ReadLine()
        _running = False
        computeThread.Join()

        Console.WriteLine("正在关闭...")
        API.API_Exit_MainThread()
        API.API_shutdown()
        Console.WriteLine("已关闭。")
    End Sub

End Module

File: Py/cross/cross_call.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
CrossCall – 并发客户端（Consumer）
功能：连接到 ipc:cross，在独立线程中交替调用 'add' 和 'inv_seri'，
持续 10 秒后自动退出。可同时启动多个实例以模拟负载。
与 Pascal cross_call 完全等价。
"""
import sys
import os
import time
import random
import threading

# 将上级目录（Py）加入模块搜索路径
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))

from api_hub import _native
from api_hub.core import DataHandle


# ========== 远程调用封装（原始二进制） ==========

def add__(a, b):
    """
    封装 'add' 远程调用。
    与 Pascal cross_call 中 add__ 函数完全等价。
    """
    with DataHandle("add") as send:
        send.write_int32(a)
        send.write_int32(b)

        res_ptr = _native.API_Call(b"demo", send.raw, 2000)
        if not res_ptr:
            print(f"[Call] add({a}, {b}) 返回空句柄")
            return 0

        size = _native.API_GetSize(res_ptr)
        if size == 0:
            _native.API_Free_DataHnd(res_ptr)
            print(f"[Call] add({a}, {b}) 超时或失败")
            return 0

        # 包装结果句柄并读取
        with DataHandle._from_raw(res_ptr, owned=True) as result:
            return result.read_int32()


def inv_seri__():
    """
    封装 'inv_seri' 远程调用。
    数据布局与 Pascal cross_call 中 inv_seri_ 函数完全等价。
    """
    # 固定测试数据（与 Pascal 完全一致）
    b = 200
    w = 0x10
    c = 0x2F
    u64 = 0x3F
    s = "hello world"
    f = 3.14

    with DataHandle("inv_seri") as send:
        send.write_uint8(b)
        send.write_uint16(w)
        send.write_uint32(c)
        send.write_uint64(u64)
        send.write_string_null_terminated(s)
        send.write_single(f)

        res_ptr = _native.API_Call(b"demo", send.raw, 2000)
        if not res_ptr:
            return "inv_seri 返回空句柄"

        size = _native.API_GetSize(res_ptr)
        if size == 0:
            _native.API_Free_DataHnd(res_ptr)
            return "inv_seri 超时或失败"

        with DataHandle._from_raw(res_ptr, owned=True) as result:
            f_ret = result.read_single()
            s_ret = result.read_string_null_terminated()
            u64_ret = result.read_uint64()
            c_ret = result.read_uint32()
            w_ret = result.read_uint16()
            b_ret = result.read_uint8()

            return (f"接收数据序 [{b_ret}, {w_ret}, {c_ret}, {u64_ret}, \"{s_ret}\", {f_ret:.2f}] = "
                    f"发送数据序 [{f_ret:.2f}, \"{s_ret}\", {u64_ret}, {c_ret}, {w_ret}, {b_ret}]")


# ========== 工作线程 ==========

RUN_DURATION = 10  # 秒

def do_compute(stop_event):
    """
    工作线程的主循环，与 Pascal cross_call 中 Do_Compute 完全等价。
    """
    start_time = time.time()
    print(f"[Call] 仿真计算启动（可多开），持续 {RUN_DURATION} 秒...")
    while not stop_event.is_set() and (time.time() - start_time) < RUN_DURATION:
        # 随机选择调用 add 或 inv_seri
        if random.choice([True, False]):
            a = random.randint(1, 2**31 - 1)
            b = random.randint(1, 2**31 - 1)
            result = add__(a, b)
            if result != 0:
                remaining = RUN_DURATION - (time.time() - start_time)
                print(f"[Call] 计算 \"a({a})+b({b})\" = 计算结果 {result} ({remaining:.2f}秒以后退出)")
        else:
            status = inv_seri__()
            remaining = RUN_DURATION - (time.time() - start_time)
            print(f"[Call] {status} ({remaining:.2f}秒退出)")

        # 与 Pascal 中 TCompute.Sleep(1) 等价
        time.sleep(0.001)


def main():
    print("=== CrossCall (Python) – Concurrent Client ===")

    # 注册 Shutdown Hook
    import atexit
    def cleanup():
        print("[Shutdown] Exiting main thread and shutting down.")
        _native.API_Exit_MainThread()
        _native.API_shutdown()
    atexit.register(cleanup)

    try:
        # ---------- 1. 连接服务（纯消费，不暴露 API） ----------
        _native.API_Reset_Prepare()
        _native.API_Prepare_Client(b"ipc:cross", None)

        if _native.API_Prepare_Done() != 1:
            print("[ERROR] prepareDone() failed. Check console output.")
            return
        print("[OK] Connected to ipc:cross")

        # ---------- 2. 启动工作线程 ----------
        stop_event = threading.Event()
        thread = threading.Thread(target=do_compute, args=(stop_event,))
        thread.daemon = True
        thread.start()

        # ---------- 3. 等待工作线程结束 ----------
        # 等待 10 秒后，设置停止标志，并等待线程真正退出
        time.sleep(RUN_DURATION + 0.5)  # 多等 0.5 秒确保最后一次循环完成
        stop_event.set()
        thread.join(timeout=2)

        print("[OK] 计算完成，清理线程中.")

    except KeyboardInterrupt:
        print("\n[INFO] Interrupted by user.")
    except Exception as e:
        print(f"[FATAL] Unexpected error: {e}")
        import traceback
        traceback.print_exc()
    finally:
        cleanup()
        print("[OK] Client shutdown complete.")


if __name__ == "__main__":
    main()